Added support for VMA

johannesugb opened the pull request:
Yo, I've added support for VMA, because I'd like to tweet to Chris Glover that he now gets what he wants in Vulkan Launchpad. ;) 

I'm hereby asking you for a quick review of the code.
Usage of VMA can be switched on/off by defining/undefining the `USE_VMA` macro.

Known issues:
- Some image does not get destroyed properly... gotta look into that later
- One cannot switch on VMA if Vulkan Launchpad is built as a library.

Regarding the latter point: There is currently a `#define USE_VMA` at the top of `VulkanLaunchpad.cpp`. But that's uncool.      
What would be cool, would be to add it before including Vulkan Launchpad, i.e., like so:            
```cpp
#define USE_VMA 
#include <VulkanLaunchpad.h>
```
buuut that doesn't work right now with Vulkan Launchpad Starter (I think) because Vulkan Launchpad is built as a library. Any suggestions about that? 

I think there is that "`INCLUDE`"-library type in CMake, which does not really build it as a library but merely includes all the source files. Would that be an option?

johannesugb committed:
Added support for VMA

johannesugb commented:
Nah, after giving it some more thought, I would like to pursue a different strategy, namely one where Vulkan Launchpad enables VMA functionality based on the presence or absence of the VMA header using [`__has_include`](https://en.cppreference.com/w/cpp/preprocessor/include).

I'll push the changes soon.

fknfilewalker commented:
Will review this soon. Just short comment, we can add VMA as a submodule then it is always present.

johannesugb commented:
> Will review this soon. Just short comment, we can add VMA as a submodule then it is always present.

I'm not sure if that is necessary since it is an integral part of the Vulkan SDK now. My plan was: If it is installed, it is usable (see my previous comment for my planned approach).

johannesugb committed:
Refactored VMA support a bit:
Based on the check `#if __has_include(<vma/vk_mem_alloc.h>)`, Vulkan Launchpad provides an additional `vklInitFramework` overload.

johannesugb commented:
Refactored VMA support (see last commit message).
Example source code how to use the refactored version can be found in the `vkl_has_vma` branch of `AnIntroductionToVulkan`

fknfilewalker commented on VulkanLaunchpad.h:
```diff
@@ -204,11 +209,21 @@ extern const char *vklRequiredInstanceExtensions[];
+ *  Initializes the framework.
+ *  This overload additionally takes a VmaAllocator handle
+ */
+bool vklInitFramework(VkInstance vk_instance, VkSurfaceKHR vk_surface, VkPhysicalDevice vk_physical_device,
```
Should the user really create the vma allocator? We can derive everything from the given inputs that is needed

johannesugb replied:
Yeah, we could... except for creating a depth buffer image, which must happen before the call to `vklInitFramework`. 🤷‍♂️ 

Therefore, I thought that it might be the best approach to:
- Let Vulkan Launchpad support VMA if it _can_ be found by the compiler.
- Make it opt-in
- The user can always use it to create resources manually, even before the call to `vklInitFramework`---otherwise, how to let the user have VMA allocate memory for the depth buffer image?

Also, letting the user setup VMA manually could be argued to serve a didactic purpose.

fknfilewalker replied:
I though vma should happen in the background and will not be noticeable by the user. So the user interaction with vma should then be kinda 50/50?

johannesugb replied:
Yeah, my first approach was 99% trsnsparent (besides a flag/switch to enable it).

But then I switched to 50/50 for the reasons stated above. What do you think?

johannesugb replied:
And you know what's cool?!
If you don't do anything, everything stays the same as before.
But if you create a VMA allocator, the you can use VMA---I think, this would make a great addition to the depth buffer task.
